gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcApiExceptionFactory.java

```diff
@@ -67,8 +84,65 @@ ApiException create(Throwable throwable) {
     }
   }
 
-  private ApiException create(Throwable throwable, Status.Code statusCode) {
+  private ApiException create(Throwable throwable, Status.Code statusCode, Metadata metadata) {
     boolean canRetry = retryableCodes.contains(GrpcStatusCode.grpcCodeToStatusCode(statusCode));
-    return ApiExceptionFactory.createException(throwable, GrpcStatusCode.of(statusCode), canRetry);
+    GrpcStatusCode grpcStatusCode = GrpcStatusCode.of(statusCode);
+
+    if (metadata == null) {
+      return ApiExceptionFactory.createException(throwable, grpcStatusCode, canRetry);
+    }
+
+    byte[] bytes = metadata.get(Metadata.Key.of(ERROR_DETAIL_KEY, Metadata.BINARY_BYTE_MARSHALLER));
+    if (bytes == null) {
+      return ApiExceptionFactory.createException(throwable, grpcStatusCode, canRetry);
+    }
+
+    com.google.rpc.Status status;
+    try {
+      status = com.google.rpc.Status.parseFrom(bytes);
+    } catch (InvalidProtocolBufferException e) {
+      return ApiExceptionFactory.createException(throwable, grpcStatusCode, canRetry);
+    }
+
+    return ApiExceptionFactory.createException(
+        throwable, grpcStatusCode, canRetry, buildErrorDetails(status));
+  }
+
+  private ErrorDetails buildErrorDetails(com.google.rpc.Status status) {
+    ErrorDetails.Builder errorDetailsBuilder = ErrorDetails.builder();
+    for (Any detail : status.getDetailsList()) {
+      addErrorDetail(errorDetailsBuilder, detail);
+    }
+    return errorDetailsBuilder.build();
+  }
+
+  private void addErrorDetail(ErrorDetails.Builder errorDetailsBuilder, Any detail) {
+    try {
+      if (detail.is(ErrorInfo.class)) {
```

Comment:
Just wondering, is it possible to convert this to a switch statement somehow? For example using class names as strings, as switch accepts strings as the switched value.

Reply:
I gave it some thoughts as well but couldn't find a good way to use switch statements, I don't think using class names as String would work either. Because we are not checking a variable, we are checking the result of method call, and the input of the method call is a variable. We can only get the type of `Any` through this method call instead of an attribute, otherwise we could've check it easily. 
Let me know if you have some other good ideas.